### PR TITLE
linux: check for required config options before building

### PR DESCRIPTION
**Pull request by lrusak**
Pretty self explanatory. There are linux config options we require when building in order for LibreELEC to run. This just provides a way to sanity check that before building. This is helpful when adding new platforms or changing kernel configs.

maybe the `required_options` file can be renamed/move to `distributions/LibreELEC` as it may be specific to us.

I didn't actually test the current platforms with this and it would be helpful if people tested this on the various platforms so we can make those changes before merging.

I've added the required options for IWD here also, I'm not sure if they should live in another location or not but this keeps it simple.

### CODE REVIEW

**Comment by heitbaum** [comment]
Tested on amlogic - needs the following change 5335734d55f7d8be380a0351efcc8e3aec07a66a

**Comment by jernejsk** [comment]
AW needs these changes: https://github.com/LibreELEC/LibreELEC.tv/commit/e98285753dc00f8ee55d933e369504531f3c74cb

Otherwise I'm fine with concept of checking for required options. Should we expand it with standard set of wifi, bluetooth, dvb and similar drivers?

**Comment by lrusak** [comment]
Thanks guys!!


> Otherwise I'm fine with concept of checking for required options. Should we expand it with standard set of wifi, bluetooth, dvb and similar drivers?

Yes I think we should try and standardize many of the main features. This will likely require some expansion of the logic in this PR to be able to check for built in options vs modules. Likely a follow up PR would be good for that.

**Comment by HiassofT** [comment]
A possibility to check kernel configs is certainly nice, but this is better suited for a separate script eg in tools that can be run independently of image builds.

Deliberately breaking builds is a very bad idea, kernel config options sometimes change and this has a huge potential of causing lots of grief when testing with newer or older kernel versions.

A linter-style script that can output errors and warnings, maybe with an option to auto-fix settings when run interactively, would be a lot more helpful.

Ideally we'd integrate that into our CI so we get notified about potential kconfig issues (eg green/yellow/orange status etc), without breaking builds because the checker script is too strict (as it's currently the case - no idea why CONFIG_CRYPTO_USER_API_HASH needs to by "y", "m" will do just fine)

**Comment by lrusak** [comment]
I think two things may be getting mixed here.

I'm really just wanting to add the required config options to make LibreELEC "LibreELEC".

I think part of this _is_ to break the build, otherwise the build won't be considered LibreELEC if that makes sense. If even one of some of these options is missing the build won't run so what's the point in building?

I think we really should standardize the kernel config now that most platforms are on the same kernel version. it shouldn't just be some optional linter that may be run if the maintainer decides to. That's just my opinion of course.

So we can either leave anything listed as required or break it up into "required to run" and "required to be LibreELEC" where the later wouldn't break the build but would log the issue.

We could also force enabling the options at build time but I think this would cause more problems as it may not be obvious and may cause shipping a build that doesn't relate to the sources.

Just my 2c.

**Comment by CvH** [comment]
Alpine does it with some common config and just config changes ontop
full config https://github.com/alpinelinux/aports/blob/master/main/linux-lts/config-lts.aarch64
device specific changes https://github.com/alpinelinux/aports/blob/master/main/linux-rpi/config-changes-rpi.armhf

**Comment by lrusak** [comment]
> Alpine does it with some common config and just config changes ontop
> 
> full config https://github.com/alpinelinux/aports/blob/master/main/linux-lts/config-lts.aarch64
> 
> device specific changes https://github.com/alpinelinux/aports/blob/master/main/linux-rpi/config-changes-rpi.armhf

@sraue tried to do something like this for openelec and I'm not really sure I like breaking up the kernel config into smaller chunks as it makes it more difficult to make changes (IMO).

It's also difficult because we still have a few platforms that have specific patch sets which may change the kernel configuration for those platforms.

I'll see what I can come up with that may be more robust than this current PR.

**Comment by lrusak** [comment]
Ok I've updated this with some changes.

I've modified the logic so we can check for built-in [y] and modules [m].

I've added a script to allow modifying kernel configs using the kernel `scripts/config` script. My only issue with this is that we may have issues if options require other options as we don't run `make olddefconfig` after changing the options. This may be something to look into.

Both utilities have to ability to check required and optional options via `required_kernel_options` and `optional_kernel_options` respectively. Optional options will just be checked but won't break the build. Required options will still break the build.

Output looks like this:
```
$ PROJECT=Generic ARCH=x86_64 ./scripts/check_kernel_config                                                                                                                                                    ✭linux-required-config
[y] CONFIG_DRM
[y] CONFIG_SQUASHFS
[y] CONFIG_SQUASHFS_ZSTD
[y] CONFIG_VFAT_FS
[y] CONFIG_EXT4_FS
[y] CONFIG_CRYPTO_USER_API_HASH
[y] CONFIG_CRYPTO_SHA1
[y] CONFIG_CRYPTO_HMAC
[y] CONFIG_CRYPTO_MD5
[y] CONFIG_CRYPTO_AES
[y] CONFIG_CRYPTO_CMAC
[y] CONFIG_CRYPTO_SHA256
[y] CONFIG_CRYPTO_SHA512
[y] CONFIG_CRYPTO_USER_API_SKCIPHER
[y] CONFIG_CRYPTO_DES
[y] CONFIG_CRYPTO_ECB
[y] CONFIG_CRYPTO_CBC
[y] CONFIG_KEY_DH_OPERATIONS
[y] CONFIG_KEYS
[y] CONFIG_ASYMMETRIC_KEY_TYPE
[y] CONFIG_ASYMMETRIC_PUBLIC_KEY_SUBTYPE
[y] CONFIG_X509_CERTIFICATE_PARSER
[y] CONFIG_PKCS7_MESSAGE_PARSER
[m] CONFIG_PKCS8_PRIVATE_KEY_PARSER
```

File: packages/linux/package.mk
Comment (knaerzche) [inline packages/linux/package.mk:179]: I guess  `kernel_make oldconfig` should be run before the options are checked, since the Kconfig scripts may change the actual options.

Comment (lrusak) [reply packages/linux/package.mk:179]: I think we should get away from running oldconfig in general as it may change options that we have configured

**Comment by lrusak** [comment]
added RPi config changes

File: packages/linux/package.mk
Comment (knaerzche) [reply packages/linux/package.mk:179]: Kconfig scripts will be invoked before the actual kernel built anyway (iirc it does even call the oldconfig target before) - they are just there to ensure options set are sane (and overwrite them if necessary). So: I'd keep it here and just check the options after make oldconfig is called to make sure, we are checking the actual options that are used for kernel build.

Comment (lrusak) [reply packages/linux/package.mk:179]: `make oldconfig` is called right after these changes

Comment (knaerzche) [reply packages/linux/package.mk:179]: looks like explained my thoughts wrong :) 
what I tried to say was: Kconf checks/corrections will be done by the kernel target itself (i.e. zImage/Image/uImage/....) anyway - no matter if we call oldconfig before or not. the good thing by calling oldconfig before is,  that we know the result here already  - and can do our checks against it

**Comment by lrusak** [comment]
objections?

**Comment by heitbaum** [comment]
Rockchip build issue

```
$ PROJECT=Rockchip ARCH=arm DEVICE=RK3399 s/build linux
BUILD      linux (target)
    TOOLCHAIN      make (auto-detect)
    INSTALL      glibc (init)
    INSTALL      arm-mem (init)
    INSTALL      libc (init)
    INSTALL      busybox (init)
    INSTALL      gcc (init)
    INSTALL      plymouth-lite (init)
    INSTALL      util-linux (init)
    INSTALL      e2fsprogs (init)
    INSTALL      dosfstools (init)
    INSTALL      terminus-font (init)
    INSTALL      open-iscsi (init)
    INSTALL      initramfs (init)
    INSTALL      initramfs (target)
LINUX: required kernel options not enabled: 
CONFIG_CRYPTO_DES=y
CONFIG_CRYPTO_CBC=y
CONFIG_KEY_DH_OPERATIONS=y
CONFIG_PKCS8_PRIVATE_KEY_PARSER=m

Please run ./scripts/check_kernel_config
FAILURE: s/build linux during pre_make_target (package.mk)
```

```
$ scripts/check_kernel_config 
CLEAN      linux
    *      Removing /home/rudi/LibreELEC.tv/build.LibreELEC-Generic.x86_64-9.80-devel/build/linux-5.10.15 ...
    *      Removing /home/rudi/LibreELEC.tv/build.LibreELEC-Generic.x86_64-9.80-devel/install_pkg/linux-5.10.15 ...
UNPACK      linux
    APPLY PATCH (common - default)      packages/linux/patches/default/linux-0000-nop-gcc-plugin.patch
patching file scripts/gcc-plugin.sh
    APPLY PATCH (common - default)      packages/linux/patches/default/linux-051-ouya_controller_support.patch
patching file drivers/hid/Kconfig
Hunk #1 succeeded at 744 (offset 85 lines).
patching file drivers/hid/Makefile
Hunk #1 succeeded at 80 (offset 9 lines).
patching file drivers/hid/hid-ids.h
Hunk #1 succeeded at 933 (offset 74 lines).
patching file drivers/hid/hid-ouya.c
patching file drivers/hid/hid-quirks.c
Hunk #1 succeeded at 543 (offset 5 lines).
    APPLY PATCH (common - default)      packages/linux/patches/default/linux-058.05-hid_sony-add_autorepeat_for_PS3_remotes.patch
patching file drivers/hid/hid-sony.c
Hunk #1 succeeded at 837 with fuzz 1 (offset -283 lines).
Hunk #2 succeeded at 2807 with fuzz 1 (offset 416 lines).
    APPLY PATCH (common - default)      packages/linux/patches/default/linux-062-imon_pad_ignore_diagonal.patch
patching file drivers/media/rc/imon.c
Hunk #1 succeeded at 1427 (offset 83 lines).
    APPLY PATCH (common - default)      packages/linux/patches/default/linux-999-no-lzma-in-x86-perf-build.patch
patching file tools/perf/Makefile.config
Hunk #1 succeeded at 55 with fuzz 2 (offset 20 lines).
    APPLY PATCH (common - default)      packages/linux/patches/default/linux-999.02-0001-pm-disable-async-suspend-resume-by-default.patch
patching file kernel/power/main.c
Hunk #1 succeeded at 95 (offset 49 lines).
[y] CONFIG_DRM
[y] CONFIG_SQUASHFS
[y] CONFIG_SQUASHFS_ZSTD
[y] CONFIG_VFAT_FS
[y] CONFIG_EXT4_FS
[y] CONFIG_CRYPTO_USER_API_HASH
[y] CONFIG_CRYPTO_SHA1
[y] CONFIG_CRYPTO_HMAC
[y] CONFIG_CRYPTO_MD5
[y] CONFIG_CRYPTO_AES
[y] CONFIG_CRYPTO_CMAC
[y] CONFIG_CRYPTO_SHA256
[y] CONFIG_CRYPTO_SHA512
[y] CONFIG_CRYPTO_USER_API_SKCIPHER
[y] CONFIG_CRYPTO_DES
[y] CONFIG_CRYPTO_ECB
[y] CONFIG_CRYPTO_CBC
[y] CONFIG_KEY_DH_OPERATIONS
[y] CONFIG_KEYS
[y] CONFIG_ASYMMETRIC_KEY_TYPE
[y] CONFIG_ASYMMETRIC_PUBLIC_KEY_SUBTYPE
[y] CONFIG_X509_CERTIFICATE_PARSER
[y] CONFIG_PKCS7_MESSAGE_PARSER
[m] CONFIG_PKCS8_PRIVATE_KEY_PARSER
```

**Comment by heitbaum** [comment]
@lrusak - please include commit 43981144dc8142a306291d4e7443179b02bc5c62 For the RK3399

**Comment by lrusak** [comment]
@heitbaum thanks, picked and pushed

**Comment by lrusak** [comment]
I've added `MERGED BLOCKED` so no one accidentally merges this with the fixup commits 😜 

I've added some extra changes:

1) `oldconfig` is now run before checking the required/optional options as suggested by @knaerzche 

2) This now can check for options that shouldn't be enabled (I have some other changes that use this to be PR'd after).

3) `${PKG_BUILD}/scripts/config` is now used to adjust the kernel config options instead of sed (long overdue)

4) `scripts/adjust_kernel_config` has been added to aid in adjusting kernel configs. This is needed because we call the kernel build with some environment variables that should also be available at config time. This also makes sure that the required kernel options are checked before running the kernel config command.
```
$ PROJECT=Generic ARCH=x86_64 ./scripts/adjust_kernel_config help
A kernel config command must be specified

./scripts/adjust_kernel_config menuconfig
```

also a list of available kernel config commands can be listed
```
$ PROJECT=Generic ARCH=x86_64 ./scripts/adjust_kernel_config commands
"commands" is not an available kernel config command

available kernel config commands are:

config
nconfig
menuconfig
xconfig
gconfig
oldconfig
localmodconfig
localyesconfig
defconfig
savedefconfig
allnoconfig
allyesconfig
allmodconfig
alldefconfig
randconfig
yes2modconfig
mod2yesconfig
listnewconfig
helpnewconfig
olddefconfig
tinyconfig
testconfig
```

This adjusts the kernel config in place rather than copying it back and for from the build folder.

I didn't add any logic to make sure the required tools are available before running the kernel config command because I don't think it's necessary. Anything that is needed by the kernel build should be built before running the script. I feel that this makes sense because you should be able to build the kernel to adjust the kernel config.

Thoughts?

**Comment by lrusak** [comment]
It would be good if we could get some more eyes on this so we can move forward with unifying some linux kernel options across all the platforms.

**Comment by lrusak** [comment]
I've removed the `MERGE BLOCKED` label as I consider this good to go now after squashing

**Review by HiassofT — APPROVED** [review]
RPi2/4 build was fine, changes look good to me

**Comment by antonlacon** [comment]
If you're looking for an optional to add, CONFIG_FUSE_FS for ntfs-3g support.

**Comment by jernejsk** [comment]
It seems that CONFIG_RT_GROUP_SCHED should be always disabled: https://github.com/systemd/systemd/blob/main/README#L116
Maybe add all systemd recommended values?

**Comment by lrusak** [comment]
> It seems that CONFIG_RT_GROUP_SCHED should be always disabled: https://github.com/systemd/systemd/blob/main/README#L116
> 
> Maybe add all systemd recommended values?

I actually have that ready as a separate PR

**Comment by lrusak** [comment]
updated new scripts to use `GPL-2.0`

**Comment by lrusak** [comment]
any other objections?

**Comment by knaerzche** [comment]
Did somebody check NXP/Samsung/Qualcomm projects already? (I'm not having a ready to use build env for either of them) At least for NXP https://github.com/LibreELEC/LibreELEC.tv/blob/master/projects/NXP/devices/iMX6/linux/linux.arm.conf#L6298 has to be set to "m" if I understand https://github.com/LibreELEC/LibreELEC.tv/pull/5044/commits/5932923b6571884f8858651cad8e8666722b797e#diff-cc0bbeebb89905d5d6850d46b2901025cebf5692e64c227ba258738cfc90c435R14-R17 correctly

For RK everything is good thanks @heitbaum and thanks for moving oldconfig before checking the options @lrusak

**Comment by HiassofT** [comment]
I've thought about it a bit more and my initial concern is still valid: this should be a quality check feature in CI, not breaking builds.

Even more so as the majority of the kernel options you check for aren't even required in current official builds - none of them use iwd.

I'd suggest changing the "die" in linux package.mk into a warning.

**Comment by knaerzche** [comment]
my 5 cents about this: I would only define the really core options we require in `required_kernel_options `- if a particular package depends on more options to enabled: it should be defined in the package and package build should fail/warn about that, not the actual kernel build.

**Comment by lrusak** [comment]
> I've thought about it a bit more and my initial concern is still valid: this should be a quality check feature in CI, not breaking builds.
> 
> Even more so as the majority of the kernel options you check for aren't even required in current official builds - none of them use iwd.

> I'd suggest changing the "die" in linux package.mk into a warning.

I've explained why this is the way it is a few times now. Unless people are forced to do it they won't make the changes. Required options are required for a reason.

> my 5 cents about this: I would only define the really core options we require in `required_kernel_options `- if a particular package depends on more options to enabled: it should be defined in the package and package build should fail/warn about that, not the actual kernel build.

meh, I don't really want to litter the build system with config files. I think gentoo will fail to build a package if it requires a certain config option but I'd rather keep everything in one location.

File: scripts/adjust_kernel_config
Comment (antonlacon) [inline scripts/adjust_kernel_config:18]: Should look like:

`AVAILABLE_KERNEL_CONFIG_COMMANDS="$(ARCH=${TARGET_KERNEL_ARCH} make -C ${PKG_BUILD} ........"`

ARCH is used by the Makefile, and on aarch64 you end up with 
```
Makefile:681: arch/aarch64/Makefile: No such file or directory
make: *** No rule to make target 'arch/aarch64/Makefile'.  Stop.
"menuconfig" is not an available kernel config command
```

I'm guessing it's been a hidden error on the mixed arch devices too.

Comment (lrusak) [reply scripts/adjust_kernel_config:18]: I don't see this behavior
```
$ PROJECT=NXP DEVICE=iMX8 ARCH=arm ./scripts/adjust_kernel_config commands
<snip>
lots of unpacking output
</snip

"commands" is not an available kernel config command

available kernel config commands are:

config
nconfig
menuconfig
xconfig
gconfig
oldconfig
localmodconfig
localyesconfig
defconfig
savedefconfig
allnoconfig
allyesconfig
allmodconfig
alldefconfig
randconfig
yes2modconfig
mod2yesconfig
listnewconfig
helpnewconfig
olddefconfig
tinyconfig
testconfig
```

Comment (antonlacon) [reply scripts/adjust_kernel_config:18]: Because "arm" is also a kernel recognized architecture, so the kernel's Makefile finds something to work with at `arch/arm/Makefile`. "aarch64" is not - it wants "arm64"

Comment (lrusak) [reply scripts/adjust_kernel_config]: I've updated it to use `kernel_make`

**Review by antonlacon — APPROVED** [review]
Overall looks good to me. I'm not a big fan of how `check_kernel_config` is working, but since nothing is using the part that bothers me most (the optional check seems to just add them to the config anyway - what makes it optional?), it can be revisited when something does use it.

Run tested w/ RPi3.aarch64.

**Comment by heitbaum** [comment]
Updates for Samsung and Qualcomm
- https://github.com/lrusak/LibreELEC.tv/pull/10
- https://github.com/heitbaum/LibreELEC.tv/commit/a90639c50cd01d4ff1cab17faafd1bf02d9729e2
- https://github.com/heitbaum/LibreELEC.tv/commit/cc89482d32ba108abbb0d3c965fad1dfcc462e26

**Comment by lrusak** [comment]
exynos and dragonboard can use further updates and there is no point in updating the kernel configs at this time.

Since the build won't fail anymore there is no reason to update the kernel configs in this PR.

This PR should be good to go as is.

**Comment by HiassofT** [comment]
@lrusak was there a specific reason for moving the kernel config setup from post_patch() to pre_make_target()? Putting that into configure_target() would seem more logical to me.

We should also keep in mind that after moving .config setup away from post_patch `scripts/unpack linux` no longer creates .config and that make_host()/makeinstall_host() will now be called without a .config in place - which seems to be OK for just installing headers but is still a bit odd

**Comment by lrusak** [comment]
> @lrusak was there a specific reason for moving the kernel config setup from post_patch() to pre_make_target()? Putting that into configure_target() would seem more logical to me.
> 

pre_make_target() existed already

> We should also keep in mind that after moving .config setup away from post_patch `scripts/unpack linux` no longer creates .config and that make_host()/makeinstall_host() will now be called without a .config in place - which seems to be OK for just installing headers but is still a bit odd

I'm not sure if this is a problem or not really. I don't know if there is any target information that goes into the headers install (I hope not)

**Comment by lrusak** [comment]
A couple comments.

Now that there is no build breakage there really is no point to have required and optional as they are all just suggestions anyways.

The reason I wanted required options to break the build is the fact that they may result in a non working build (either not booting at all or core features, wireless, etc not working). I still think it's correct to have the build break as that would be desired for me rather than a not working build (and the debugging that may go along with that).

What do I do? Do I remove the optional file and just a have a single kernel_options file? Do I leave them separate for some reason?

This is now going on a month doing this PR. It's getting a bit long winded.

**Comment by lrusak** [comment]
bueller?

**Comment by lrusak** [comment]
updated to remove required/optional and just have a list of kernel options.

moved the two added scripts from the scripts folder to tools (as they should be run manually)

**Comment by lrusak** [comment]
merge?

**Comment by lrusak** [comment]
merge???

**Review by CvH — APPROVED** [review]

**Pull request merged**
